### PR TITLE
Use Reduce for Global Pool

### PR DESCRIPTION
**Pull request by wransom-TT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/28108

### Problem description
avg_pool2d used the sliding window path for all cases, including global average pooling where the kernel covers the entire spatial input. This is inefficient, global pooling is a simple reduction that doesn't need sliding window overhead. A separate global_avg_pool2d C++ device op existed that used the efficient reduction path, but having two device ops for the same operation creates maintenance burden and API confusion. In PyTorch, global pooling has no separate op, adaptive_avg_pool2d(x, (1,1)) and avg_pool2d with appropriate kernel size both route through the same pool2d kernel. We want to replicate that behavior.

### What's changed
Reduction path in avg_pool2d: When avg_pool2d detects global pooling (kernel >= input spatial dims, zero padding), it now uses pool_sum reduction instead of the sliding window path. pool_sum applies sum + scalar division in a single kernel call. For the common batch_size==1 case, this is a direct pool_sum on the input with no reshapes, just 1 device op. Multi-batch inputs are reshaped from (1,1,N*H*W,C) to (N,H,W,C) to establish per-batch structure before reducing.                                                              
                                                                  
Removed global_avg_pool2d C++ device op: Deleted the standalone C++ implementation, nanobind binding, header, and build entries. Replaced with a thin Python wrapper in pool.py that reshapes the input and delegates to avg_pool2d. The wrapper handles rank 2, 3, and 4 inputs, matching the old C++ op's API. Existing callers (ttnn.global_avg_pool2d(x)) continue to work. adaptive_avg_pool2d with output_size=(1,1) also benefits since it calls avg_pool2d internally.                                                                                                                                                                    
                                                                  
New test cases were also added.

### Checklist
Sanity APC:
https://github.com/tenstorrent/tt-metal/actions/runs/23874507585
unrelated failure

Blackhole APC:
https://github.com/tenstorrent/tt-metal/actions/runs/23874517021
unrelated failure

Nightly L2:
https://github.com/tenstorrent/tt-metal/actions/runs/23874534257

Frequent Model:
https://github.com/tenstorrent/tt-metal/actions/runs/23915233664/job/69748672793
unrelated failures

Model Perf:
https://github.com/tenstorrent/tt-metal/actions/runs/23961586102

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp [hunk 1]
```diff
@@ -981,6 +983,46 @@ static std::vector<Tensor> pool2d(
...
+
+        // Flatten spatial dims: (N, H, W, C) -> (N, 1, H*W, C)
+        auto ns = nhwc_input.padded_shape();
+        ttnn::Shape flat_shape({ns[0], 1, ns[1] * ns[2], ns[3]});
+        Tensor flat_input = ttnn::experimental::view(nhwc_input, flat_shape);
+
+        float scalar = divisor_override.has_value() ? 1.0f / float(divisor_override.value()) : 1.0f / float(hw);
+
+        Tensor output = ttnn::operations::reduction::pool_sum(
+            flat_input, int(flat_shape.rank() - 2), mem_config, compute_kernel_config, scalar);
```
[the remark this answers is not in the document]
Comment (wransom-TT) [reply ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp]: comment added

File: ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp [hunk 2]
```diff
@@ -981,6 +983,66 @@ static std::vector<Tensor> pool2d(
...
+            input_tensor_4d = ttnn::to_layout(input_tensor_4d, Layout::ROW_MAJOR);
+        }
+        auto in_shape = input_tensor_4d.padded_shape();
+
+        // reshape (1,1,N*H*W,C) → (N,H,W,C) → view (N,1,H*W,C) → reduce → reshape
+        ttnn::Shape nhwc_logical({batch_size, input_h, input_w, channels});
+        ttnn::Shape nhwc_padded({batch_size, input_h, input_w, in_shape[3]});
+        Tensor nhwc_input = ttnn::reshape(input_tensor_4d, nhwc_logical, nhwc_padded);
+
+        auto ns = nhwc_input.padded_shape();
```
Comment (dstoiljkovicTT) [inline ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp]: Let's rename this ns variable, to something a bit more verbouse

File: ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp [hunk 3]
```diff
@@ -981,6 +983,66 @@ static std::vector<Tensor> pool2d(
...
 
+    // Use reduction path for global average pooling (kernel covers entire input with no padding)
+    // This is more efficient than the sliding window path for this special case
+    std::array<uint32_t, 4> padding_check = sliding_window::get_pair_n4_padding(padding);
+    bool is_global_pool =
+        (kernel_size[0] >= input_h && kernel_size[1] >= input_w) &&
+        (padding_check[0] == 0 && padding_check[1] == 0 && padding_check[2] == 0 && padding_check[3] == 0);
+
+    if (is_global_pool && pool_type == Pool2DType::AVG_POOL2D) {
+        auto mem_config = memory_config.value_or(input_tensor_4d.memory_config());
```
Comment (dstoiljkovicTT) [inline ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp]: Maybe we can avoid using auto with the non trivial data types

File: ttnn/ttnn/operations/pool.py [hunk 4]
```diff
@@ -49,11 +49,58 @@ def golden_maxpool2d(
 ttnn.attach_golden_function(ttnn.max_pool2d, golden_maxpool2d)
 
 
+def global_avg_pool2d(input_tensor, *, memory_config=None, dtype=None):
```
Comment (dstoiljkovicTT) [inline ttnn/ttnn/operations/pool.py]: Should this be removed? I can see that there is another golden_global_avg_pool2d below

File: ttnn/ttnn/operations/pool.py [hunk 5]
```diff
@@ -49,11 +49,58 @@ def golden_maxpool2d(
...
+    )
+    if rank == 4:
+        return ttnn.reshape(result, (N, 1, 1, C))
+    elif rank == 3:
+        return ttnn.reshape(result, (1, 1, C))
+    else:
+        return ttnn.reshape(result, (1, 1))
+
+
+ttnn.global_avg_pool2d = global_avg_pool2d
```
Comment (dstoiljkovicTT) [inline ttnn/ttnn/operations/pool.py]: Not sure what this is, is this registration of the op? Should this be done this way?

File: ttnn/ttnn/operations/pool.py [hunk 6]
```diff
@@ -49,11 +49,58 @@ def golden_maxpool2d(
...
+
+
 def golden_global_avg_pool2d(input_tensor: ttnn.Tensor):
     import torch
 
     output_size = (1, 1)
-    return torch.nn.functional.global_avg_pool2d(input_tensor, output_size)
+    return torch.nn.functional.adaptive_avg_pool2d(input_tensor, output_size)
```
Comment (dstoiljkovicTT) [inline ttnn/ttnn/operations/pool.py]: We are missing a reshape here, ttnn input shape is n,h,w,c and adaptive accepts n,c,h,w maybe call a golden adaptive instead or just add a reshape here

File: ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp [hunk 7]
```diff
@@ -981,6 +983,66 @@ static std::vector<Tensor> pool2d(
...
+        uint32_t hw = input_h * input_w;
+        float scalar = divisor_override.has_value() ? 1.0f / float(divisor_override.value()) : 1.0f / float(hw);
```
Comment (dstoiljkovicTT) [inline ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp:1018]: What happens with the count include pad?

File: ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp [hunk 8]
```diff
@@ -981,6 +983,66 @@ static std::vector<Tensor> pool2d(
     }
     // For rank-4, input_tensor_4d is already the input_tensor, no copy needed
 
+    // Use reduction path for global average pooling (kernel covers entire input with no padding)
+    // This is more efficient than the sliding window path for this special case
+    std::array<uint32_t, 4> padding_check = sliding_window::get_pair_n4_padding(padding);
+    bool is_global_pool =
```
Comment (dstoiljkovicTT) [inline ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp:1004]: I think we are missing the check whether dilation is the default value to conclude it i sglobal avg pool

**Review by dstoiljkovicTT — COMMENTED** [review]
Can you please link a passing run of the Frequent model tests since this op is used in multiple models

**Review by dgomezTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp (hunk 2 above)
Comment (wransom-TT) [reply ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp]: done

File: ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp (hunk 3 above)
Comment (wransom-TT) [reply ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp]: done

File: ttnn/ttnn/operations/pool.py (hunk 4 above)
Comment (wransom-TT) [reply ttnn/ttnn/operations/pool.py]: I don't think so, the two functions are for different things. `global_avg_pool2d` is the op wrapper, and `golden_global_avg_pool2d` is the torch golden reference. Before I don't think we had the python wrapper because there was a cpp binding but that's been deleted.

File: ttnn/ttnn/operations/pool.py (hunk 5 above)
Comment (wransom-TT) [reply ttnn/ttnn/operations/pool.py]: Oh yeah see my response above

File: ttnn/ttnn/operations/pool.py (hunk 6 above)
Comment (wransom-TT) [reply ttnn/ttnn/operations/pool.py]: fixed with a permute, let me know if this looks good

File: ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp (hunk 7 above)
Comment (wransom-TT) [reply ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp:1018]: I think for the global pool path to be activated we check that padding is zero, so I don't think count include pad will change the behavior.

File: ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp (hunk 8 above)
Comment (wransom-TT) [reply ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp:1004]: fixed!

File: ttnn/ttnn/operations/pool.py (hunk 4 above)
Comment (dstoiljkovicTT) [reply ttnn/ttnn/operations/pool.py]: Oh I see this is due to the fact that pytorch doesn't have global_avg_pool op

File: ttnn/cpp/ttnn-nanobind/__init__.cpp [hunk 9]
```diff
@@ -63,7 +63,7 @@
 #include "ttnn/operations/normalization/normalization_nanobind.hpp"
 #include "ttnn/operations/point_to_point/point_to_point_nanobind.hpp"
 #include "ttnn/operations/pool/generic/generic_pools_nanobind.hpp"
-#include "ttnn/operations/pool/global_avg_pool/global_avg_pool_nanobind.hpp"
+
```
Comment (dstoiljkovicTT) [inline ttnn/cpp/ttnn-nanobind/__init__.cpp]: We have an extra line here

**Review by dstoiljkovicTT — APPROVED** [review]
This looks good to me, approved assuming Frequent models pipeline is passing
